File: 2024/day-03/main2.lua
-- Set the file to read input from
local file = arg[1] or 'input.txt'
local part2 = 0
local enabled = true

-- Function to read the file content
local function readFile(filename)
    local file = io.open(filename, "r")
    if not file then
        print("Error reading file:", filename)
        return nil
    end
    local content = file:read("*a")
    file:close()
    return content
end

-- Read the input data from the file
local D = readFile(file)
if not D then return end

-- Loop through the string
for i = 1, #D do
    -- Check for "do()"
    if D:sub(i, i+3) == "do()" then
        enabled = true
    end

    -- Check for "don't()"
    if D:sub(i, i+6) == "don't()" then
        enabled = false
    end

    -- Look for occurrences of "mul("
    if D:sub(i, i+3) == "mul(" then
        -- Find the closing parenthesis
        local j = i + 3
        while j <= #D and D:sub(j, j) ~= ")" do
            j = j + 1
        end

        if j <= #D and D:sub(j, j) == ")" then
            -- Extract numbers inside "mul(x, y)"
            local mul_str = D:sub(i+4, j-1)
            local x, y = mul_str:match("(%d+),%s*(%d+)")

            if x and y then
                x, y = tonumber(x), tonumber(y)

                -- Make sure the closing parenthesis isn't followed by a number
                if j + 1 <= #D and not D:sub(j+1, j+1):match("%d") then
                    -- Perform the multiplication if enabled
                    if enabled then
                        part2 = part2 + x * y
                    end
                    print(D:sub(i, j))  -- Print the found "mul(x, y)"
                end
            end
        end
    end
end

-- Output the final result
print(part2)

File: 2024/day-03/main2.go
package main

import (
	"fmt"
	"os"
	"regexp"
	"strconv"
	"strings"
)

func main() {
	// Check for command line argument to specify the input file
	file := "input.txt"
	if len(os.Args) > 1 {
		file = os.Args[1]
	}

	// Read input file
	data, err := os.ReadFile(file)
	if err != nil {
		fmt.Println("Error reading file:", err)
		return
	}

	// Initialize variables
	D := string(data)
	part2 := 0
	enabled := true

	// Iterate through the string
	for i := 0; i < len(D); i++ {
		// Check for "do()"
		if i+4 <= len(D) && D[i:i+4] == "do()" {
			enabled = true
		}
		// Check for "don't()"
		if i+7 <= len(D) && D[i:i+7] == "don't()" {
			enabled = false
		}

		// Look for occurrences of "mul("
		if i+4 <= len(D) && D[i:i+4] == "mul(" {
			// Find the closing parenthesis
			j := i + 4
			for j < len(D) && D[j] != ')' {
				j++
			}
			// Extract numbers inside "mul(x, y)"
			if j < len(D) && D[j] == ')' {
				// Use regex to find integers in the range
				re := regexp.MustCompile(`\d+`)
				matches := re.FindAllString(D[i:j+1], -1)

				if len(matches) == 2 {
					x, err1 := strconv.Atoi(matches[0])
					y, err2 := strconv.Atoi(matches[1])

					if err1 == nil && err2 == nil {
						// Make sure the closing parenthesis isn't followed by a number
						if j+1 >= len(D) || !strings.ContainsAny(string(D[j+1]), "0123456789") {
							// Perform the multiplication if enabled
							if enabled {
								part2 += x * y
							}
							fmt.Println(D[i : j+1]) // Print the found "mul(x, y)"
						}
					}
				}
			}
		}
	}

	// Output the final result
	fmt.Println(part2)
}
